File: java/coding_dojo/src/main/java/coding_dojo/RomanNumberAssembler.java
package coding_dojo;

import java.util.*;

class RomanNumberAssembler {
    private Map<Integer, String> arabicToRomanMap = new LinkedHashMap<>();

    RomanNumberAssembler() {
        this.arabicToRomanMap.put(1000, "M");
        this.arabicToRomanMap.put(500, "D");
        this.arabicToRomanMap.put(100, "C");
        this.arabicToRomanMap.put(50, "L");
        this.arabicToRomanMap.put(10, "X");
        this.arabicToRomanMap.put(5, "V");
        this.arabicToRomanMap.put(1, "I");
    }

    String convertToRomanNumber(Integer number) throws Exception {
        if (this.arabicToRomanMap.containsKey(number)) {
            return this.arabicToRomanMap.get(number);
        }

        StringBuilder sb = new StringBuilder();

        if (number <= 0 || number >= 5000) {
            throw new Exception("Can not be converted !!!");
        }

        while (number > 0) {
            Integer divisor = this.getDivisor(number);

            sb.append(this.arabicToRomanMap.get(divisor));

            number -= divisor;
        }

        return this.normalize(sb.toString());
    }

    private Integer getDivisor(Integer number) {
        List<Integer> list = new ArrayList<>(this.arabicToRomanMap.keySet());

//        int value = 0;
//
//        for (Integer primitive : list)
//            if (number / primitive > 0) {
//                value = primitive;
//                break;
//            }
//
//        return value;

        for (Integer primitive : list)
            if (number / primitive > 0) {
                return primitive;
            }

        return 0;
    }

    private String normalize(String number) {

        List<String> list = new ArrayList<>(this.arabicToRomanMap.values());

        for (int i = 1; i < list.size() - 1; i = i + 2) {
        /*
            ("DCCCC", "CM")
            ("CCCC", "CD")
            ("LXXXX", "XC")
            ("XXXX", "XL")
            ("VIIII", "IX")
            ("IIII", "IV")
        */
            String high = list.get(i);
            String low = list.get(i + 1);
            String highLord = list.get(i - 1);
            String low_repeated = String.join("", Collections.nCopies(4, low));

            number = number.replace(high + low_repeated, low + highLord);
            number = number.replace(low_repeated, low + high);
        }

        return number;
    }
}

File: python/coding_dojo/RomanNumberAssembler.py
class RomanNumberAssembler:
    mapping = {
        1000: "M",
        500: "D",
        100: "C",
        50: "L",
        10: "X",
        5: "V",
        1: "I"
    }

    def convert(self, number):
        roman_mapping = self.mapping.get(number)
        if roman_mapping is not None:
            return roman_mapping

        roman_number = ""
        while number > 0:
            divisor, roman = self.get_divisor(number)
            roman_number = roman_number + roman
            number -= divisor

        return self.normalize(roman_number)

    def get_divisor(self, number):
        for k, v in self.mapping.items():
            if number / k >= 1:
                return k, v

    def normalize(self, roman_number):
        values = list(self.mapping.values())

        for i in range(2, len(values), 2):
            low = values[i]
            repeated_low = low * 4
            high = values[i - 1]
            high_lord = values[i - 2]
            roman_number = roman_number.replace(high + repeated_low, low + high_lord).replace(repeated_low, low + high)

        return roman_number
